Define extension schema policy and registry (#415)
quicwg/qlog: draft-ietf-quic-qlog-main-schema.md

QlogFile = {
    qlog_version: text
    ? qlog_format: text .default "JSON"

>     ? additional_event_schema: [+ text]

Author: `schemas` seems most correct, will fix.

While we are here, I wonder if CDDL has a URI type?

>     ? additional_event_schema: [+ text]

Author: Instead of `text` is there a `uri` or somthinh

>     ? additional_event_schema: [+ text]

Author: Thanks for the clarification

~~~
{: #simulation-marker-def title="SimulationMarker definition"}

# Event extensibility {#event-extensibility}

The main qlog schema defined by this document describes log and trace formats
that have related events. A minimal set of common events are presented in
{{events}}. It is expected that logging is extended with specific, per-protocol
event definitions that specify the name (category + type) and data needed for
each individual event. Examples include the QUIC event definitions {{QLOG-QUIC}}
and HTTP/3 event definitions {{QLOG-H3}}.

New event definitions SHOULD follow the guidance in this section, which provides
consistency that makes it easier for qlog implementers to support a wide variety
of protocols.

## Extended Event Schema

> New events SHOULD be defined using an extension schema, using the annotations

Author: Thought a bit more on this. What I think we want is for logical groups of qlog events to be defined under a single schema. There should not be a restriction that a single IETF document defines only a single schema - for example, we might want to have a single document that defines events for a bunch of QUIC extensions and split them out by extension-specific schemas.

I don't think the current text, or the proposal, articulate this well. Will try to come up with a some different text.

> New events SHOULD be defined using an extension schema, using the annotations

Reviewer: See new comment on this below :) you changed this to "event definitions" which is better, but still a bit vague.

At least the example currently makes clear that a single document can contain multiple schema, and that event types should belong to categories, which in turn should belong to a schema. 

In short: I still think this need clarification.

and concepts presented in this section.

Each extension schema is named by a URI. That URI MUST be absolute; it precisely
identifies the format and meaning of the extension. URIs that contain a domain
name SHOULD also contain a month-date in the form mmyyyy. The definition of the

> element and assignment of the URI MUST have been authorized by the owner of the

Reviewer: unclear what "element" is referring to specifically here...

Author: nor me, I copy-pasted it from RFC 8285. Will change.

domain name on or very close to that date. (This avoids problems when domain
names change ownership.) If the resource or document defines several categories

> or groups of events, then the URI MUST identify the actual extension in use,

Reviewer: this could do with an example maybe? not sure what that would look like if you're for example using quic events... would you need to list #connectivity and #transport as separate subentries in `additional_event_schema` or not? Where do you draw the line?

Reviewer: Is there an option to say: if you don't define # or ?, then you mean EVERYTHING included in the URI. If you want to select only specific parts, use # or ? to specify that? 

e.g., urn:ietf:params:qlog:quic means ALL QUIC events, while urn:ietf:params:qlog:quic#recovery means ONLY the recovery events and nothing else? How does RFC 8285 manage such things?

Author: This is an open question we should ask in the meeting

Author: I spoke with some people in the hallways. There's a reason that RFC 8285 does it that way, due to how RTP stuff works. However, everyone seemed to agree that requiring the full expansions wasn't that onerous and was more clear. Allowing the url without those seems like a microoptimization.

There's another reason the full expansion is nice. If any documents in the future extend a namespace, then its not clear what extensions are used or not. That makes the implicit include quite insipid.

e.g., using a fragment or query identifier (characters after a "#" or "?" in the
URI).

For extensions defined in RFCs, the URI used SHOULD be a URN starting with

> `urn:ietf:params:qlog` followed by a registered, descriptive name.

Reviewer: ```suggestion
`urn:ietf:params:qlog:` followed by a registered, descriptive name.
```

Prefix should end with `:` no?

A log file that that uses events from extension schema SHOULD list all schema
identifiers in the `additional_event_schema` field.

For example, a qlog file that uses two hypothetical qlog extension schema named
"rick" and "morty" would indicate this as:

~~~
additional_event_schema = [
                            urn:ietf:params:qlog:rick,
                            urn:ietf:params:qlog:morty

>                           ]

Reviewer: This could do with an extension of the example to also show at least one domainname-based extension (i.e., things that are not defined in RFCs), with the mmyyyy date in the URI

~~~
{: #simulation-marker-def title="SimulationMarker definition"}

# Event extensibility {#event-extensibility}

The main qlog schema defined by this document describes log and trace formats
that have related events. A minimal set of common events are presented in
{{events}}. It is expected that logging is extended with specific, per-protocol
event definitions that specify the name (category + type) and data needed for
each individual event. Examples include the QUIC event definitions {{QLOG-QUIC}}
and HTTP/3 event definitions {{QLOG-H3}}.

New event definitions SHOULD follow the guidance in this section, which provides
consistency that makes it easier for qlog implementers to support a wide variety
of protocols.

## Extended Event Schema

New event definitions SHOULD be part of an extension schema, using the
annotations and concepts presented in this section.

> A schema defines one or more qlog event categories that can contain multiple

Reviewer: ```suggestion
An extension schema defines one or more qlog event categories that can contain multiple
```

event definitions. Each event MUST only belong to a single category.

> It is not possible to extend a category with new events. Instead, extension

Reviewer: ```suggestion
It is not possible to extend an existing category, defined in a different document, with new events. Instead, extension
```

Author: I don't think we need the "different document" qualification, there may be no document at all for something like a private extension. The blanket rule is always the same, no extension.

> schema MUST define new categories with globally unique identifiers. Per

Reviewer: ```suggestion
schemas MUST define new categories with globally unique identifiers. Per
```

{{{#name-field}}}, event names are the concatenation of category and type.
Therefore, unique category identifiers ensure events in qlog files are
disambuiguated.

Each extension schema is named by a URI. That URI MUST be absolute; it precisely

> identifies the format and meaning of the extension. URIs that contain a domain

Reviewer: Not exactly sure what the intention is of including "format" here... might be confused with "serialization format", which is certainly not the intent. Replace with a different word or remove?

Author: lol, still a hangover from copy/paste. What we should say here is that a schema URI precisely identifies the category name and the event type(s) (the CDDL definition)

name SHOULD also contain a month-date in the form mmyyyy. The definition of the
schema and assignment of the URI MUST have been authorized by the owner of the
domain name on or very close to that date. (This avoids problems when domain

> names change ownership.) If the resource or document defines several event

Reviewer: ```suggestion
names change ownership.) If the extension schema contains several event
```

categories, then the URI MUST identify the actual category in use, e.g., using

> a fragment or query identifier (characters after a "#" or "?" in the URI).

Reviewer: I wonder why we'd use an "e.g.," here... are there other accepted options? if not, remove the e.g.,. If yes, I'd like to see at least one in the example :)

Author: This was also a copy/paste. I don't see a good argument for allowing flexibility to be honest. How about we just restrict it to `#`?

Author: done

For extensions defined in RFCs, the URI used SHOULD be a URN starting with
`urn:ietf:params:qlog:` followed by a registered, descriptive name.

> A log file that uses events from extension schema SHOULD list all schema

Reviewer: ```suggestion
A log file that uses events from extension schemas SHOULD list all extension schema
```

> identifiers in the `additional_event_schemas` field; see {{qlog-file-schema}}

Reviewer: ```suggestion
identifiers (potentially multiple per extension schema in the case multiple categories are used) in the `additional_event_schemas` field; see {{qlog-file-schema}}
```

I feel that at this point, there's lots of ambiguity between a "schema identifier" and a specification of it for a specific category and what should actually be listed in the `additional_event_schemas`. The example below makes it a bit clearer, but I think we need more in this prose as well. I think the above is the minimum we could get away with, but potentially more text is needed (and maybe a new term like "category identifier"?)

Author: See new text

> and {{qlog-file-seq-schema}}.

Reviewer: So here, I'm wondering if we should give guidance to tools on how to actually use the `additional_event_schemas` field? Like, they SHOULD use it to see if they support the listed event categories, but they SHOULD NOT bork on unexpected events (and rather just ignore them). 

put differently: I think we need to make clear that tools/users MUST NOT treat this as a fully qualified list and expect all/none events to be present, but rather more as a hint to further improve operations.

Author: Yeah that would be good to have.

Author: See new text

Reviewer: I think the new text sufficiently covers this indeed.

> In the following example, a qlog file contains two hypothetical standardized

Reviewer: ```suggestion
In the following example, a qlog file contains events defined in two hypothetical standardized
```

extension schema, along with a private extension. The first schema is named
"rick" and consists of categories "roll", "astley", and "moranis". The second
schema is named "john" and consists of categories "doe", "candy", and "goodman".
The private schema is named "pickle" and consists of categories "pepper",
"lilly", and "rick":

~~~

> additional_event_schemas = [
>                             urn:ietf:params:qlog:rick#roll,
>                             urn:ietf:params:qlog:rick#astley,
>                             urn:ietf:params:qlog:rick#moranis,
>                             urn:ietf:params:qlog:john#doe,
>                             urn:ietf:params:qlog:john#candy,
>                             urn:ietf:params:qlog:john#goodman,
>                             https://example.com/032024/pickle.html#pepper,
>                             https://example.com/032024/pickle.html#lilly,
>                             https://example.com/032024/pickle.html#rick,
>                           ]

Reviewer: ```suggestion
"additional_event_schemas": [
                            "urn:ietf:params:qlog:rick#roll",
                            "urn:ietf:params:qlog:rick#astley",
                            "urn:ietf:params:qlog:rick#moranis",
                            "urn:ietf:params:qlog:john#doe",
                            "urn:ietf:params:qlog:john#candy",
                            "urn:ietf:params:qlog:john#goodman",
                            "https://example.com/032024/pickle.html#pepper",
                            "https://example.com/032024/pickle.html#lilly",
                            "https://example.com/032024/pickle.html#rick",
                          ]
```

Since it's an example of a JSON serialized qlog file, it should be valid JSON.

> ~~~

Reviewer: ```suggestion
~~~
{: #additional-event-schema-ex title="Example of using the additional-event-schema field"}
```

~~~
{: #simulation-marker-def title="SimulationMarker definition"}

# Event extensibility {#event-extensibility}

The main qlog schema defined by this document describes log and trace formats

> that have related events. A minimal set of common events are presented in

Reviewer: ```suggestion
that have related events. A minimal set of common events is presented in
```

{{events}}. It is expected that logging is extended with specific, per-protocol
event definitions that specify the name (category + type) and data needed for
each individual event. Examples include the QUIC event definitions {{QLOG-QUIC}}
and HTTP/3 event definitions {{QLOG-H3}}.

New event definitions SHOULD follow the guidance in this section, which provides

> consistency that makes it easier for qlog implementers to support a wide variety

Reviewer: ```suggestion
consistency measures that make it easier for qlog implementers to support a wide variety
```

of protocols.

## Extended Event Schema

New event definitions SHOULD be part of an extension schema, using the
annotations and concepts presented in this section.

An extension schema defines one or more qlog event categories that can contain

> multiple event definitions. Each category MUST have a globally unique category

Reviewer: we are being ambiguous as to what exactly an "event definition" is.

Above, you define it as `per-protocol
event definitions that specify the name (category + type) and data` while here, it's more correct to say each category contains multiple event types... in the next sentence then, you switch to just the word "event".

It's not a major issue, but I feel we need consistent naming for these things...

For that, I would say `event type` here instead of `event definition` and in the next sentence `Each event type MUST only belong` instead of `Each event MUST`

Related: I think it would really help to have an illustrative example of this at the start of this part

identifier. Each event MUST only belong to a single category. Per
{{{#name-field}}}, event names are the concatenation of category and type.
Therefore, these requirements ensure there are no ambiguities when multiple

> extension schema are in use.

Reviewer: I understand what you're saying because I have context, and it's an important thing to highlight, BUT it's at the wrong place in the text... new readers don't know why this is important because you haven't talked about the URIs yet (and how event names tie into those).

Author: This section is all about how to extend qlog with more concrete event types. The URI stuff is ancilliary to that.

We need to start with clear requirements on what extensions can and can't do. 

Events have a name that conveys the category and type. When event names are not unique, ambiguitiy can cause tooling to fail. The language here is normative to avoid that class of problem. This problem exists even if we weren't trying to invent a way to indicate what an endpoint was generating e.g. - if we both invent a foobar:frame_sent event type with wildly different data formats.

Reviewer: I agree that we have to have clear requirements. I'm mainly saying that this specific text either needs to move (original suggestion) or needs to be explained better in-place (which seems your suggestion) to really make sense for the reader.

I'm guessing this will be easily solved if we clearly explain this up front:

schema namespace > category identifier > event type 

and then it's easier to use that terminology to explain this restriction.

Implementations that might record events from extension schemas SHOULD list all
category identifiers in use. This is achieved by including the appropriate URI
in the `additional_event_schemas` field of the QlogFile ({{qlog-file-schema}})
or QlogFileSeq ({{qlog-file-seq-schema}}). The `additional_event_schema` is a

> hint to tools about the possible events that a qlog file might contain. The file

Reviewer: ```suggestion
hint to tools about the possible event categories (and event types contained therein) that a qlog file might contain. The file
```

> may contain events that do not belong to a listed category identifier. Tools

Reviewer: ```suggestion
may contain event types that do not belong to a listed category identifier. Tools
```

MUST NOT treat this as an error; see {{tooling}}.

Each extension schema is named by a URI. That URI MUST be absolute and MUST

> include a category identifier, indicated using a fragment identifier (characters

Reviewer: This is inconsistent with how you defined a "schema" at the start. There you say:

> An extension schema defines one or more qlog event categories

and here 1 schema is named by 1 URI, but that URI then somehow contains a category identifier? So the URI doesn't identify the schema, it instead identifies a single category WITHIN an extension schema.

Author: agree, are you happier with something like 

```suggestion
Each category identifier in a schema is named by a URI. That URI MUST be absolute and MUST contain a fragment identifier (characters after a "#" in the URI) that indicates the category identifier. 
```

Reviewer: yes, that's much more accurate :)

after a "#" in the URI).

For extension schema defined in RFCs, the URI used SHOULD be a URN starting with
`urn:ietf:params:qlog:` followed by a registered, descriptive name.

> Private or non-standard extension schema a can use other URI formats. URIs that

Reviewer: ```suggestion
Private or non-standard extension schema can use other URI formats. URIs that
```

contain a domain name SHOULD also contain a month-date in the form mmyyyy. The
definition of the schema and assignment of the URI MUST have been authorized by
the owner of the domain name on or very close to that date. (This avoids
problems when domain names change ownership.)

In the following example, a qlog file contains events defined in two hypothetical standardized

> extension schema, along with a private extension. The first schema is named

Reviewer: it's a bit confusing that here we have 3 schema in a single file... while not impossible, most files will have a single schema (i.e., quic and h3).

I would rework this to saying that we have a single file with 2 schema (rick and john) and another file with a private extension schema (pickle).

Author: I don't think its confusing. For the HTTP/3 qlog I'd generate for quiche, I'm already expecting to do the following based on identifiers in the QUIC and H3 extension schema

```
~~~
"additional_event_schemas": [
                            "urn:ietf:params:qlog:quic#quic",
                            "urn:ietf:params:qlog:quic#security",
                            "urn:ietf:params:qlog:quic#recovery",
                            "urn:ietf:params:qlog:h3#h3"
                          ]
~~~

"rick" and consists of categories "roll", "astley", and "moranis". The second
schema is named "john" and consists of categories "doe", "candy", and "goodman".
The private schema is named "pickle" and consists of categories "pepper",

> "lilly", and "rick":

Reviewer: As I said above, I think it would REALLY help to have this example (or part of it) at the start of this section, as it makes it much clearer what you mean by schema and category

Author: That's specification via example, which is bad. The language needs to be clear enough on its own other wise people only code to the example.

Reviewer: I agree that wouldn't be ideal, but then indeed we need to have better text imo :) 

Maybe something more generic up top, like:

``` urn:ietf:params:qlog:<SCHEMA-NAMESPACE>#<CATEGORY-IDENTIFIER> 

...

For example, this document defines the "main" schema namespace with two category identifiers: "generic" and "simulation".

``` 

to help explain how the concepts map to the URIs.

(see {{raw-info}}). Therefore, qlog users should exercise caution and limit the
inclusion of such fields for all but the most stringent use cases.

# IANA Considerations {#iana}

IANA is requested to register a new entry in the "IETF URN Sub-namespace for
Registered Protocol Parameter Identifiers" registry ({{!RFC3553}})":

Registered Parameter Identifier:
: qlog

Reference:
: This Document

IANA Registry Reference:
: [](https://www.iana.org/assignments/qlog){: brackets="angle"}

IANA is requested to create the "qlog event extension schema identifer" registry
at [](https://www.iana.org/assignments/qlog) for the purpose of registering
event extension schema. It has the following format:

> | Extension URI | Categor Identifier(s) | Description | Reference |

Reviewer: ```suggestion
| Extension URI | Category Identifier(s) | Description | Reference |
```

Thanks to Jana Iyengar, Brian Trammell, Dmitri Tikhonov, Stephen Petrides, Jari
Arkko, Marcus Ihlar, Victor Vasiliev, Mirja Kühlewind, Jeremy Lainé, Kazu

> Yamamoto, Christian Huitema, Hugo Landau, and Jonathan Lennow for their feedback

Reviewer: ```suggestion
Yamamoto, Christian Huitema, Hugo Landau, and Jonathan Lennox for their feedback
```
